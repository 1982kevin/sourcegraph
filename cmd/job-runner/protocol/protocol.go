package protocol

// Request is a request made to the job-runner service.
type Request struct {
	// TODO: the commands we would run
	// TODO: a way to pass build context (repository)
	// TODO: a way to pass timeout (or just have the remote end hangup?)
}

// Response is one of many responses the job-runner service streams back to the
// requestor. It does this by keeping the HTTP connection open and sending back
// newline-delimited JSON responses with this type (see https://ndjson.org).
type Response struct {
	// TODO: stderr/stdout logs
	// TODO: exit code of process
	// TODO: a way to signal job timeout
}
