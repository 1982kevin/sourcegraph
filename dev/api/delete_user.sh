#!/usr/bin/env bash

set -e
unset CDPATH
cd "$(dirname "${BASH_SOURCE[0]}")/../.."

eval "$(grep 'export OVERRIDE_AUTH_SECRET=' dev/start.sh)"

cd ./dev/api

username_to_delete="$1"

user_id_to_delete_data=$(jq --arg username "$username_to_delete" '.variables={"username":$username}' <user_id_query.json)

user_id_to_delete=$(
  curl -sS \
    -H "X-Override-Auth-Secret: $OVERRIDE_AUTH_SECRET" \
    -H 'X-Override-Auth-Username: dev' \
    -H 'Content-Type: application/json; charset=utf-8' \
    -XPOST \
    -d "$user_id_to_delete_data" \
    http://localhost:3080/.api/graphql |
    jq -c -r '.data.user.id'
)

if [ "$user_id_to_delete" = "null" ]; then
  echo User not found: "$username_to_delete"
  exit 1
fi

delete_user_data=$(jq --arg user "$user_id_to_delete" '.variables={"user":$user}' <delete_user_mutation.json)

curl -sS \
  -H "X-Override-Auth-Secret: $OVERRIDE_AUTH_SECRET" \
  -H 'X-Override-Auth-Username: dev' \
  -H 'Content-Type: application/json; charset=utf-8' \
  -XPOST \
  -d "$delete_user_data" \
  http://localhost:3080/.api/graphql
echo
