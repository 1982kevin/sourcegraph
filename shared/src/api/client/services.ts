import { PlatformContext } from '../../platform/context'
import { ReferenceParams } from '../protocol'
import { createContextService } from './context/contextService'
import { CommandRegistry } from './services/command'
import { ContributionRegistry } from './services/contribution'
import { LinkPreviewProviderRegistry } from './services/linkPreview'
import { TextDocumentLocationProviderIDRegistry, TextDocumentLocationProviderRegistry } from './services/location'
import { NotificationsService } from './services/notifications'
import { PanelViewProviderRegistry } from './services/panelViews'
import { createViewService } from './services/viewService'
import { createWorkspaceService } from './services/workspaceService'

/**
 * Services is a container for all services used by the client application.
 */
export class Services {
    constructor(
        private platformContext: Pick<
            PlatformContext,
            | 'settings'
            | 'updateSettings'
            | 'requestGraphQL'
            | 'getScriptURLForExtension'
            | 'clientApplication'
            | 'sideloadedExtensionURL'
            | 'createExtensionsService'
        >
    ) {}

    public readonly commands = new CommandRegistry()
    public readonly context = createContextService(this.platformContext)
    public readonly workspace = createWorkspaceService()
    // public readonly model = createModelService()
    // public readonly viewer = createViewerService(this.model)
    public readonly notifications = new NotificationsService()
    public readonly contribution = new ContributionRegistry(
        this.viewer,
        this.model,
        this.platformContext.settings,
        this.context.data
    )
    public readonly linkPreviews = new LinkPreviewProviderRegistry()
    public readonly textDocumentDefinition = new TextDocumentLocationProviderRegistry()
    public readonly textDocumentReferences = new TextDocumentLocationProviderRegistry<ReferenceParams>()
    public readonly textDocumentLocations = new TextDocumentLocationProviderIDRegistry()
    public readonly panelViews = new PanelViewProviderRegistry()
    public readonly view = createViewService()
}
