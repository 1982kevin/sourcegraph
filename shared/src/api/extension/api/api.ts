import { ProxyMarked } from 'comlink'
import { InitData } from '../extensionHost'
import { ExtensionWindowsAPI } from './windows'
import { FlatExtensionHostAPI } from '../../contract'

export type ExtensionHostAPIFactory = (initData: InitData) => ExtensionHostAPI

export interface ExtensionHostAPI extends ProxyMarked, FlatExtensionHostAPI {
    ping(): 'pong'

    windows: ExtensionWindowsAPI
}
