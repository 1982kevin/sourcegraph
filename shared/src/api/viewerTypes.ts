import { Selection } from '@sourcegraph/extension-api-types'

export type Viewer = CodeEditor | DirectoryViewer
export type ViewerData = CodeEditorData | DirectoryViewerData

/**
 * ViewerId exposes the unique ID of a viewer.
 */
export interface ViewerId {
    /** The unique ID of the viewer. */
    readonly viewerId: string
}

export interface BaseViewerData {
    readonly isActive: boolean
}

export interface DirectoryViewerData extends BaseViewerData {
    readonly type: 'DirectoryViewer'
    /** The URI of the directory that this viewer is displaying. */
    readonly resource: string
}

/**
 * Describes a code viewer to be created.
 */
export interface CodeEditorData extends BaseViewerData {
    readonly type: 'CodeEditor'

    /** The URI of the model that this viewer is displaying. */
    readonly resource: string

    readonly selections: Selection[]
}

/**
 * Describes a code editor that has been added to the {@link ViewerService}.
 */
export interface CodeEditor extends ViewerId, CodeEditorData {}

/**
 * Describes a directory viewer that has been added to the {@link ViewerService}.
 */
export interface DirectoryViewer extends ViewerId, DirectoryViewerData {}

export type ViewerWithPartialModel = CodeEditorWithPartialModel | DirectoryViewer // Directories don't have a model

/**
 * A code editor with a partial model.
 *
 * To get the editor's full model, look up the model in the {@link ModelService}.
 */
export interface CodeEditorWithPartialModel extends CodeEditor {
    model: PartialModel
}

/**
 * A code editor with its full model, including the model text.
 */
export interface CodeEditorWithModel extends CodeEditor {
    /** The code editor's model. */
    model: TextModel
}

export type ViewerUpdate =
    | ({ type: 'added'; viewerData: ViewerData } & ViewerId)
    | ({ type: 'updated'; viewerData: Pick<CodeEditorData, 'selections'> } & ViewerId)
    | ({ type: 'deleted' } & ViewerId)
