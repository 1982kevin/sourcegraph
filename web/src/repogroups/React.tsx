import { RepogroupMetadata } from './types'
import { SearchPatternType } from '../../../shared/src/graphql/schema'
import * as React from 'react'
export const react: RepogroupMetadata = {
    title: 'React',
    name: 'react',
    url: '/react',
    description:
        "Explore React repositories on GitHub. The search examples show how to use React Hook useState with various data types.",
    examples: [
        {
            title: 'Find imports of useState with regex search',
            exampleQuery: <>import [^;]+useState[^;]+ from 'react'</>,
            rawQuery: "import [^;]+useState[^;]+ from 'react'",
            patternType: SearchPatternType.regexp,
        },
        {
            title: 'Examples of useState with an object as the input parameter',
            description: `useState takes a single argument and can accept a primitive, an array or an object.
            Syntax for usage is [state_value, function_to_update_state_value] = useState (initial_state_value).`,
            exampleQuery: <>{'useState({:[string]})'}</>,
            rawQuery: 'useState({:[string]}) count:1000',
            patternType: SearchPatternType.structural,
        },
        {
            title: 'Examples of useState with any JavaScript data type as the input parameter',
            exampleQuery: <>useState(:[string])</>,
            rawQuery: 'useState(:[string]) count:1000',
            patternType: SearchPatternType.structural,
        },
        {
            title: 'Examples of useState with any JavaScript data type as the input parameter in only TypeScript files',
            exampleQuery: (
                <>
                    useState(:[string])
                    <span className="search-keyword">lang:</span>typescript
                </>
            ),
            rawQuery: 'useState(:[string]) count:1000 lang:typescript',
            patternType: SearchPatternType.structural,
        },
    ],
    homepageDescription: 'Learn React with code examples.',
    homepageIcon: 'https://cdn4.iconfinder.com/data/icons/logos-3/600/React.js_logo-512.png',
}
