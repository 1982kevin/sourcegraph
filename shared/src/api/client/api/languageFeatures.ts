import { Remote, ProxyMarked, proxyMarker } from 'comlink'
import { Location } from '@sourcegraph/extension-api-types'
import { DocumentSelector, Unsubscribable } from 'sourcegraph'
import { ProxySubscribable } from '../../extension/api/common'
import { ReferenceParams, TextDocumentPositionParams } from '../../protocol'
import { TextDocumentLocationProviderIDRegistry, TextDocumentLocationProviderRegistry } from '../services/location'
import { registerRemoteProvider } from './common'

/** @internal */
export interface ClientLanguageFeaturesAPI extends ProxyMarked {
    $registerDefinitionProvider(
        selector: DocumentSelector,
        providerFunction: Remote<((params: TextDocumentPositionParams) => ProxySubscribable<Location[]>) & ProxyMarked>
    ): Unsubscribable & ProxyMarked
    $registerReferenceProvider(
        selector: DocumentSelector,
        providerFunction: Remote<((params: ReferenceParams) => ProxySubscribable<Location[]>) & ProxyMarked>
    ): Unsubscribable & ProxyMarked

    /**
     * @param idStr The `id` argument in the extension's {@link sourcegraph.languages.registerLocationProvider}
     * call.
     */
    $registerLocationProvider(
        idStr: string,
        selector: DocumentSelector,
        providerFunction: Remote<((params: TextDocumentPositionParams) => ProxySubscribable<Location[]>) & ProxyMarked>
    ): Unsubscribable & ProxyMarked
}

/** @internal */
export class ClientLanguageFeatures implements ClientLanguageFeaturesAPI, ProxyMarked {
    public readonly [proxyMarker] = true

    constructor(
        private definitionRegistry: TextDocumentLocationProviderRegistry,
        private referencesRegistry: TextDocumentLocationProviderRegistry<ReferenceParams>,
        private locationRegistry: TextDocumentLocationProviderIDRegistry
    ) {}

    public $registerDefinitionProvider(
        documentSelector: DocumentSelector,
        providerFunction: Remote<((params: TextDocumentPositionParams) => ProxySubscribable<Location[]>) & ProxyMarked>
    ): Unsubscribable & ProxyMarked {
        return registerRemoteProvider(this.definitionRegistry, { documentSelector }, providerFunction)
    }

    public $registerReferenceProvider(
        documentSelector: DocumentSelector,
        providerFunction: Remote<((params: TextDocumentPositionParams) => ProxySubscribable<Location[]>) & ProxyMarked>
    ): Unsubscribable & ProxyMarked {
        return registerRemoteProvider(this.referencesRegistry, { documentSelector }, providerFunction)
    }

    public $registerLocationProvider(
        id: string,
        documentSelector: DocumentSelector,
        providerFunction: Remote<((params: TextDocumentPositionParams) => ProxySubscribable<Location[]>) & ProxyMarked>
    ): Unsubscribable & ProxyMarked {
        return registerRemoteProvider(this.locationRegistry, { id, documentSelector }, providerFunction)
    }
}
