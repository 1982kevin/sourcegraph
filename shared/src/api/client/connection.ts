import * as comlink from 'comlink'
import { from, Subject, Subscription } from 'rxjs'
import { first } from 'rxjs/operators'
import { ContextValues, Progress, ProgressOptions, Unsubscribable } from 'sourcegraph'
import { PlatformContext, ClosableEndpointPair } from '../../platform/context'
import { ExtensionHostAPIFactory } from '../extension/api/api'
import { InitData } from '../extension/extensionHost'
import { ClientAPI } from './api/api'
import { createClientContent } from './api/content'
import { ClientContext } from './api/context'
import { ClientLanguageFeatures } from './api/languageFeatures'
import { ClientViews } from './api/views'
import { ClientWindows } from './api/windows'
import { Services } from './services'
import {
    MessageActionItem,
    ShowInputParams,
    ShowMessageRequestParams,
    ShowNotificationParams,
} from './services/notifications'
import { registerComlinkTransferHandlers } from '../util'
import { initMainThreadAPI } from './mainthread-api'
import { isSettingsValid } from '../../settings/settings'
import { FlatExtHostAPI } from '../contract'

export interface ExtensionHostClientConnection {
    /**
     * Closes the connection to and terminates the extension host.
     */
    unsubscribe(): void
}

/**
 * An activated extension.
 */
export interface ActivatedExtension {
    /**
     * The extension's extension ID (which uniquely identifies it among all activated extensions).
     */
    id: string

    /**
     * Deactivate the extension (by calling its "deactivate" function, if any).
     */
    deactivate(): void | Promise<void>
}

/**
 * @param endpoints The Worker object to communicate with
 */
export async function createExtensionHostClientConnection(
    endpointsPromise: Promise<ClosableEndpointPair>,
    services: Services,
    initData: Omit<InitData, 'initialSettings'>,
    platformContext: Pick<PlatformContext, 'settings' | 'updateSettings'>
): Promise<{ subscription: Unsubscribable; api: comlink.Remote<FlatExtHostAPI> }> {
    const subscription = new Subscription()

    // MAIN THREAD

    registerComlinkTransferHandlers()

    const endpoints = await endpointsPromise
    subscription.add(endpoints.subscription)

    /** Proxy to the exposed extension host API */
    const initializeExtensionHost = comlink.wrap<ExtensionHostAPIFactory>(endpoints.proxy)

    const initialSettings = await from(platformContext.settings).pipe(first()).toPromise()
    const proxy = await initializeExtensionHost({
        ...initData,
        // TODO what to do in error case?
        initialSettings: isSettingsValid(initialSettings) ? initialSettings : { final: {}, subjects: [] },
    })

    const clientContext = new ClientContext((updates: ContextValues) => services.context.updateContext(updates))
    subscription.add(clientContext)

    const clientWindows = new ClientWindows(
        (parameters: ShowNotificationParams) => services.notifications.showMessages.next({ ...parameters }),
        (parameters: ShowMessageRequestParams) =>
            new Promise<MessageActionItem | null>(resolve => {
                services.notifications.showMessageRequests.next({ ...parameters, resolve })
            }),
        (parameters: ShowInputParams) =>
            new Promise<string | null>(resolve => {
                services.notifications.showInputs.next({ ...parameters, resolve })
            }),
        ({ title }: ProgressOptions) => {
            const reporter = new Subject<Progress>()
            services.notifications.progresses.next({ title, progress: reporter.asObservable() })
            return reporter
        }
    )

    const clientViews = new ClientViews(
        services.panelViews,
        services.textDocumentLocations,
        services.viewer,
        services.view
    )

    const clientLanguageFeatures = new ClientLanguageFeatures(
        services.textDocumentDefinition,
        services.textDocumentReferences,
        services.textDocumentLocations,
        services.completionItems
    )

    const clientContent = createClientContent(services.linkPreviews)

    const { api: newAPI, subscription: apiSubscriptions } = initMainThreadAPI(proxy, platformContext, services)

    subscription.add(apiSubscriptions)

    const clientAPI: ClientAPI = {
        ping: () => 'pong',
        context: clientContext,
        languageFeatures: clientLanguageFeatures,
        windows: clientWindows,
        views: clientViews,
        content: clientContent,
        ...newAPI,
    }

    comlink.expose(clientAPI, endpoints.expose)

    return { subscription, api: proxy }
}
