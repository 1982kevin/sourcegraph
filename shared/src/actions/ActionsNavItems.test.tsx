import * as H from 'history'
import React from 'react'
import { NOOP_TELEMETRY_SERVICE } from '../telemetry/telemetryService'
import { ActionsNavItems } from './ActionsNavItems'
import { ContributableMenu } from '../api/protocol'
import { of, NEVER } from 'rxjs'
import { Services } from '../api/client/services'
import { mount } from 'enzyme'

jest.mock('mdi-react/OpenInNewIcon', () => 'OpenInNewIcon')

describe('ActionItem', () => {
    const MOCK_EXTENSIONS_CONTROLLER = { executeCommand: () => Promise.resolve(undefined) }
    const NOOP_PLATFORM_CONTEXT = { forceUpdateTooltip: () => undefined, settings: NEVER }
    const location = H.createLocation(
        'https://github.com/sourcegraph/sourcegraph/pull/5287/files#diff-eb9883bb910397a210512a13fd7384ac'
    )

    test('Renders contributed action items', () => {
        const component = mount(
            <ActionsNavItems
                menu={ContributableMenu.EditorTitle}
                location={location}
                extensionsController={{
                    ...MOCK_EXTENSIONS_CONTROLLER,
                    services: {
                        contribution: {
                            getContributions: () =>
                                of({
                                    actions: [
                                        {
                                            id: 'a',
                                            actionItem: {
                                                label: 'Action A',
                                                description: 'This is Action A',
                                            },
                                        },
                                    ],
                                    menus: {
                                        'editor/title': [
                                            {
                                                action: 'a',
                                            },
                                        ],
                                    },
                                }),
                        },
                    } as Services,
                }}
                platformContext={NOOP_PLATFORM_CONTEXT}
                telemetryService={NOOP_TELEMETRY_SERVICE}
            />
        )
        expect(component.children()).toMatchSnapshot()
    })
})
