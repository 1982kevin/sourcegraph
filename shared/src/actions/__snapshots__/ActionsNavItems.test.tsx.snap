// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionItem Renders contributed action items 1`] = `
Array [
  " ",
  <li>
    <ActionItem
      action={
        Object {
          "actionItem": Object {
            "description": "This is Action A",
            "label": "Action A",
          },
          "id": "a",
        }
      }
      className="actions-nav-items__action-item"
      extensionsController={
        Object {
          "executeCommand": [Function],
          "services": Object {
            "contribution": Object {
              "getContributions": [Function],
            },
          },
        }
      }
      key="a"
      location="[Location path=https://github.com/sourcegraph/sourcegraph/pull/5287/files#diff-eb9883bb910397a210512a13fd7384ac]"
      menu="editor/title"
      platformContext={
        Object {
          "forceUpdateTooltip": [Function],
          "settings": "[Observable]",
        }
      }
      telemetryService={
        Object {
          "log": [Function],
        }
      }
      variant="actionItem"
    >
      <span
        className="action-item actions-nav-items__action-item action-item--variant-action-item"
        data-tooltip="This is Action A"
      >
         
        Action A
      </span>
    </ActionItem>
  </li>,
]
`;
